providers/azure_test.go

```diff
@@ -122,82 +122,134 @@ func testAzureBackend(payload string) *httptest.Server {
 	assert.Equal(t, nil, err)
-	assert.Equal(t, "user@windows.net", email)
+	assert.Equal(t, "user@windows.net", session.Email)
 }
 
-func TestAzureProviderGetEmailAddressGetUserPrincipalName(t *testing.T) {
+func TestAzureProviderEnrichSessionStateGetUserPrincipalName(t *testing.T) {
 	b := testAzureBackend(`{ "mail": null, "otherMails": [], "userPrincipalName": "user@windows.net" }`)
 	defer b.Close()
 
 	bURL, _ := url.Parse(b.URL)
 	p := testAzureProvider(bURL.Host)
 
 	session := CreateAuthorizedSession()
-	email, err := p.GetEmailAddress(context.Background(), session)
+	err := p.EnrichSessionState(context.Background(), session)
 	assert.Equal(t, nil, err)
-	assert.Equal(t, "user@windows.net", email)
+	assert.Equal(t, "user@windows.net", session.Email)
 }
 
-func TestAzureProviderGetEmailAddressFailToGetEmailAddress(t *testing.T) {
+func TestAzureProviderEnrichSessionStateFailToEnrichSessionState(t *testing.T) {
 	b := testAzureBackend(`{ "mail": null, "otherMails": [], "userPrincipalName": null }`)
 	defer b.Close()
 
 	bURL, _ := url.Parse(b.URL)
 	p := testAzureProvider(bURL.Host)
 
 	session := CreateAuthorizedSession()
-	email, err := p.GetEmailAddress(context.Background(), session)
+	err := p.EnrichSessionState(context.Background(), session)
 	assert.Equal(t, "type assertion to string failed", err.Error())
-	assert.Equal(t, "", email)
+	assert.Equal(t, "", session.Email)
 }
 
-func TestAzureProviderGetEmailAddressEmptyUserPrincipalName(t *testing.T) {
+func TestAzureProviderEnrichSessionStateEmptyUserPrincipalName(t *testing.T) {
 	b := testAzureBackend(`{ "mail": null, "otherMails": [], "userPrincipalName": "" }`)
 	defer b.Close()
 
 	bURL, _ := url.Parse(b.URL)
 	p := testAzureProvider(bURL.Host)
 
 	session := CreateAuthorizedSession()
-	email, err := p.GetEmailAddress(context.Background(), session)
+	err := p.EnrichSessionState(context.Background(), session)
 	assert.Equal(t, nil, err)
-	assert.Equal(t, "", email)
+	assert.Equal(t, "", session.Email)
 }
 
-func TestAzureProviderGetEmailAddressIncorrectOtherMails(t *testing.T) {
+func TestAzureProviderEnrichSessionStateIncorrectOtherMails(t *testing.T) {
 	b := testAzureBackend(`{ "mail": null, "otherMails": "", "userPrincipalName": null }`)
 	defer b.Close()
 
 	bURL, _ := url.Parse(b.URL)
 	p := testAzureProvider(bURL.Host)
 
 	session := CreateAuthorizedSession()
-	email, err := p.GetEmailAddress(context.Background(), session)
+	err := p.EnrichSessionState(context.Background(), session)
 	assert.Equal(t, "type assertion to string failed", err.Error())
-	assert.Equal(t, "", email)
+	assert.Equal(t, "", session.Email)
+}
+
+func TestAzureProviderEnrichSessionStateFromIDToken(t *testing.T) {
+	p := testAzureProvider("")
+
+	session := CreateAuthorizedSession()
+	session.IDToken = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwiZW1haWwiOiJmb29AYmFyLmNvbSIsImlhdCI6MTUxNjIzOTAyMn0.XRuL4Y2VPSToNB8vMvmlB-X3BwahUJzUXNx6vmzODjk"
```

Comment:
Can you add comments that nicely display the inner decoded JWT payload you are using in these tests so they are easier for the next contributors to eyeball and understand what is being tested?

Reply:
sure